cleanlab/rank.py

```diff
@@ -0,0 +1,159 @@
+# Copyright (C) 2017-2022  Cleanlab Inc.
+# This file is part of cleanlab.
+#
+# cleanlab is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero General Public License as published
+# by the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# cleanlab is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with cleanlab.  If not, see <https://www.gnu.org/licenses/>.
+
+
+# ## Rank
+#
+# Methods for ordering data.
+# This includes ranking data by most confusing to a model, most likely to be an
+# error or label issue, and related types of sorting / ordering of data.
+import numpy as np
+
+
+def order_label_issues(
+    label_issues_mask,
+    labels,
+    psx,
+    rank_by="normalized_margin",
+):
+    """Sorts label issues by normalized margin.
+    See https://arxiv.org/pdf/1810.05369.pdf (eqn 2.2)
+    e.g. normalized_margin = prob_label - max_prob_not_label
+
+    Parameters
+    ----------
+    label_issues_mask : np.array (bool)
+      Contains True if the index of labels is an error, o.w. false
+
+    labels : np.array
+      A binary vector of labels, which may contain label issues.
+
+    psx : np.array (shape (N, K))
+      P(labels=k|x) is a matrix with K probabilities for all N examples x.
+      This is the probability distribution over all K classes, for each
+      example, regarding whether the example has label labels==k P(labels=k|x). psx
+      should computed using 3 (or higher) fold cross-validation.
+
+    rank_by : str ['normalized_margin', 'prob_given_label']
+      Method to order label error indices (instead of a bool mask), either:
+        'normalized_margin' := normalized margin (p(label = k) - max(p(label != k)))
+        'prob_given_label' := [psx[i][labels[i]] for i in label_issues_idx]
+
+    Returns
+    -------
+      label_issues_idx : np.array (int)
+        Return the index integers of the label issues, ordered by
+        the normalized margin."""
+
+    assert (len(psx) == len(labels))
+    # Convert bool mask to index mask
+    label_issues_idx = np.arange(len(labels))[label_issues_mask]
+    # self-confidence is the holdout probability that an example belongs to its given class label
+    self_confidence = np.array(
```

Comment:
Can reuse get_self_confidence_for_each_label() function below to get self_confidence.

Reply:
Good point. Rewrote entire function so now its easily extensible to any new `label quality score` ranking function.

e.g.

```python
    assert (len(pred_probs) == len(labels))
    # Convert bool mask to index mask
    label_issues_idx = np.arange(len(labels))[label_issues_mask]
    pred_probs_er, labels_er = pred_probs[label_issues_mask], labels[label_issues_mask]
    if rank_by == "self_confidence":
        label_quality_scores = get_self_confidence_for_each_label(labels_er, pred_probs_er)    
    elif rank_by == "normalized_margin":        
        label_quality_scores = get_normalized_margin_for_each_label(labels_er, pred_probs_er)
    else:
        raise ValueError(
            'rank_by must be "self_confidence" or "normalized_margin", '
            'but is "' + rank_by + '".'
        )
    return label_issues_idx[np.argsort(label_quality_scores)]
```